pre_commit/languages/docker.py

```diff
@@ -78,7 +79,11 @@ def install_environment(
 
 def get_docker_user() -> Tuple[str, ...]:  # pragma: win32 no cover
     try:
-        return ('-u', f'{os.getuid()}:{os.getgid()}')
+        ver_cmd = 'docker', '--version'
+        if subprocess.check_output(ver_cmd).startswith(b'podman version '):
```

Comment:
I'm still fairly certain this is a _bug_ in podman -- also we can't incur this overhead repeatedly so this patch as written won't be accepted

at most `('docker', '--version')` should be called once per `pre-commit` run (instead of every call) -- you can use `@functools.lru_cache(maxsize=1)` to do something like that

there's also a `cmd_output` helper which does some improved things over `subprocess.check_call` which is preferable here

the coverage failure is because your new code is untested

is there a link to the podman issue?

Reply:
Sorry @asottile, I am not a developer, I reached my limits of copy-paste programming. I am trying my best. This not working is currently a blocker to start using the tool on some project, but I have only tech writer skills.

The podman issue is there: https://github.com/containers/libpod/issues/6431 
The -u option exists in podman, but in my case, the UID of my user is too high, so that change of UID won't work.

The whole story is, there is no point using this `-u` option when using podman, because it is the default behaviour of podman to map to the current user. As I understand it, this `-u` option is used here to overcome the default behaviour of docker which is not doing this mapping. Therefore, it makes more sense not adding this option when running with podman, than trying to make it work for no benefits.

Reply:
With the risk of upsetting @themr0c I have support @asottile here, that is not a bug in pre-commit itself, more of a side effect of faking `docker` cli with another container engine.

Still, reading this change I realized that the way is currently implemented this code will fail to run with remote docker hosts, where all execution runs on another host. In that case there is no mapping to perform because a  volume mount is not even possible.

AFAIK, the only way to make this work in a portable way would be to build the source code inside the container and run it. An approach without any mounts would also avoid this issue, and likely work with podman too.

Reply:
No worries for upsetting me @ssbarnea :) What I am looking for is a solution, not my solution :) 

As a Fedora 32 user, to use pre-commit I need to make pre-commit and podman interact in a friendly way. All what I have now the the `podman-docker` package. On Fedora 32 installing docker is not an option. See https://docs.docker.com/engine/install/fedora/. 

As an user with id 105971, I would also like the -u option behave friendly with me of course. Given that comment: https://github.com/containers/libpod/issues/6431#issuecomment-636119379, I would like to reproduce on docker, but ...

Given your last remark, I wonder how pre-commit will run on kubernetes. It is our next step:  we need to make it work on https://che.openshift.io.

Reply:
New commit attempting to detect a rootless environment. Is that better?

Reply:
@asottile is this cacheable?